app/src/ui/create-tag/create-tag-dialog.tsx

```diff
@@ -0,0 +1,184 @@
+  readonly dispatcher: Dispatcher
+  readonly onDismissed: () => void
+  readonly targetCommitSha: string
+  readonly initialName?: string
+}
+
+interface ICreateTagState {
+  readonly currentError: Error | null
+  readonly proposedName: string
+  readonly sanitizedName: string
+
+  /**
+   * Note: once tag creation has been initiated this value stays at true
+   * and will never revert to being false. If the tag creation operation
+   * fails this dialog will still be dismissed and an error dialog will be
+   * shown in its place.
+   */
+  readonly isCreatingTag: boolean
+
+  readonly localTags: Set<string>
+}
+
+const MAX_TAG_NAME_LENGTH = 245
+
+/** The Create Tag component. */
+export class CreateTag extends React.Component<
+  ICreateTagProps,
+  ICreateTagState
+> {
+  public constructor(props: ICreateTagProps) {
+    super(props)
+
+    this.state = {
+      currentError: null,
+      proposedName: props.initialName || '',
+      sanitizedName: '',
+      isCreatingTag: false,
+      localTags: new Set(),
+    }
+  }
+
+  public async componentDidMount() {
+    if (this.state.proposedName.length > 0) {
+      this.updateTagName(this.state.proposedName)
+    }
+
+    // Get the existing tags so we can warn the user that the chosen tag already
+    // exists before submitting.
+    // Since this is just an UX improvement, we don't need to block the rendering
+    // of the dialog (or show any loader) while we get the tags.
+    const localTags = await this.props.dispatcher.getAllTags(
+      this.props.repository
+    )
+
+    this.setState({
+      localTags: new Set(localTags),
+    })
+  }
+
+  public render() {
+    const disabled =
+      this.state.proposedName.length <= 0 ||
+      !!this.state.currentError ||
+      /^\s*$/.test(this.state.sanitizedName)
+    const error = this.state.currentError
+
+    return (
+      <Dialog
+        id="create-tag"
+        title={__DARWIN__ ? 'Create a Tag' : 'Create a tag'}
+        onSubmit={this.createTag}
+        onDismissed={this.props.onDismissed}
+        loading={this.state.isCreatingTag}
+        disabled={this.state.isCreatingTag}
+      >
+        {error ? <DialogError>{error.message}</DialogError> : null}
+
+        <DialogContent>
+          <Row>
+            <TextBox
+              label="Name"
+              value={this.state.proposedName}
+              autoFocus={true}
+              onValueChanged={this.updateTagName}
+            />
+          </Row>
+
+          {this.renderTagNameWarning()}
+        </DialogContent>
+
+        <DialogFooter>
+          <OkCancelButtonGroup
+            okButtonText={__DARWIN__ ? 'Create Tag' : 'Create tag'}
+            okButtonDisabled={disabled}
+          />
+        </DialogFooter>
+      </Dialog>
+    )
+  }
+
+  private renderTagNameWarning() {
+    const proposedName = this.state.proposedName
+    const sanitizedName = this.state.sanitizedName
+
+    if (proposedName.length > 0 && /^\s*$/.test(sanitizedName)) {
```

Comment:
I like the distinction @outofambit is making between blocking vs. non blocking - I think we should use warnings when we're not blocking you from proceeding, and errors when you are blocked.

Reply:
> should this case be an error (and covered in getCurrentError) instead of a warning? Seems like its a blocker to making the tag. (again, i know this doesn't match the create branch dialog 😞). @ampinsk do you have thoughts here?

Oh yes! As you assumed I blindly copied this from the create branch dialog 😅I agree that this should be an error since it blocks the tag creation.

I'm going to also change this in the branch creation dialog in a separate PR.

Reply:
> It looks like its to make sure there's no leading whitespace in the sanitizedName, is that right?

The regexp makes sure that the sanitized tag name does not consist only of empty spaces (or an empty string). Based on the sanitization logic, it doesn't look like the sanitized tag name can ever consist of empty characters, so I think that checking only that `!== ''` is enough. Gonna change that.